### PR TITLE
[Data Discovery] Add data layer and generalize async row fetching

### PR DESCRIPTION
**Pull request by tfrcarvalho**
# Description

This PR is the first step to make project and visualizations load data paginated data on-demand without redundant code
* Add a new data layer to hold and manage DD data 
* Move function that loads sample rows to the data layer and generalize

# Notes

* There should be no functionality changes.

# Tests

To test this change, we need to check the table behavior as we scroll or set any conditions that influence the rows to fetch (i.e. filters, search and project selection)/

### CODE REVIEW

File: app/assets/src/components/views/discovery/discovery_data_layer.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      : stopIndex;
+    let missingIdxs = range(startIndex, minStopIndex + 1);
+    if (objects.order) {
+      missingIdxs = missingIdxs.filter(
+        idx => !(objects.order[idx] in objects.entries)
+      );
+    }
+    if (missingIdxs.length > 0) {
+      // currently loads using limit and offset
+      // could eventually lead to redundant fetches if data is not requests in regular continuous chunks
```
Comment (jshoe) [inline app/assets/src/components/views/discovery/discovery_data_layer.js]: nit: s/requests/requested

File: app/assets/src/components/views/discovery/DiscoveryDataLayer.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    };
+
+    this.apiFunctions = {
+      samples: this.fetchSamples,
+    };
+  }
+
+  newObjectDb = () => ({
+    entries: {},
+    order: null,
```
Comment (jshoe) [inline app/assets/src/components/views/discovery/DiscoveryDataLayer.js]: `order` seems a little vague (at first I was expecting it to be a sorting function) ?

Comment (tfrcarvalho) [reply app/assets/src/components/views/discovery/DiscoveryDataLayer.js]: orderedIds?

Comment (jshoe) [reply app/assets/src/components/views/discovery/DiscoveryDataLayer.js]: yeah that sounds good

**Pull request merged**